### DISCUSSION TITLE
[Feature Suggestion] Budget

### DISCUSSION THREAD

**Discussion by logeshpaul** [ananthakumaran/paisa#8] [Ideas]
Income and Expense tracking are so great in this app, Kudos!

If we can consider adding budget, then it will help a lot.

**Comment by JavierMNieto** [comment]
For setting budgeting goals, I like hledger's report shown below for showing how much you have gone over or under budget.

<pre><div class="buttons"><button class="fa fa-copy clip-button" title="Copy to clipboard" aria-label="Copy to clipboard"><i class="tooltiptext"></i></button></div><code class="language-shell hljs"><span class="hljs-meta">$</span><span class="bash"> hledger balance -f bcexample.hledger -f budget.journal -MB -b 2013 --budget cur:USD</span>
Balance changes in 2013/01/01-2014/10/31:

                          ||                            2013/01                            2013/02                             2013/03 
==========================++===========================================================================================================
 &lt;unbudgeted&gt;:Expenses    ||                        5980.60 USD                        3997.35 USD                         4941.63 USD 
 &lt;unbudgeted&gt;:Liabilities ||                         293.09 USD                        -147.51 USD                          -66.01 USD 
 Assets:US                ||      1893.32 USD [26% of 7335 USD]      2929.77 USD [40% of 7335 USD]     -3898.89 USD [-53% of 7335 USD] 
 Expenses:Food            ||        396.46 USD [79% of 500 USD]        481.48 USD [96% of 500 USD]        603.32 USD [121% of 500 USD] 
 Expenses:Health          ||       290.70 USD [145% of 200 USD]        193.80 USD [97% of 200 USD]         193.80 USD [97% of 200 USD] 
 Expenses:Home            ||     2544.98 USD [100% of 2545 USD]     2545.02 USD [100% of 2545 USD]      2544.97 USD [100% of 2545 USD] 
 Expenses:Transport       ||       120.00 USD [100% of 120 USD]       120.00 USD [100% of 120 USD]        120.00 USD [100% of 120 USD] 
 Income:US                || -15119.10 USD [141% of -10700 USD]  -10331.21 USD [97% of -10700 USD]  -11079.40 USD [104% of -10700 USD] 
--------------------------++-----------------------------------------------------------------------------------------------------------
                          ||                       -3599.95 USD                        -211.30 USD                        -6640.58 USD 

</code></pre>

Having a visualization of this would be great. At first glance, this seems most appropriate for the monthly expenses page. I imagine a second bar appearing for categories that have set budgets. Below I tried to show a second bar for food where 5,000 has been budgeted for every month. There could then be text to show specifically how much of a percentage of your budget you spent. Here, 4,700 was spent of the 5,000 (94%).

<img width="938" alt="image" src="https://github.com/ananthakumaran/paisa/assets/22333691/4bcbf47d-cd02-4958-a7c5-21dfa799ba6a">

Maybe something similar can be done for the other bar graphs for expenses. I don't see it being very useful to see in the calendar or postings sections unless for forecasted transactions.

**Reply by ananthakumaran** [reply]
Hi @JavierMNieto, I added a new page for the budget instead of changing the expense page. You can find more details in the [docs](https://ananthakumaran.in/paisa/budget.html). I am unsure if I can show all the budget-related information on the monthly expense page.

PS: Somehow I didn't get any notification for your comment and just saw this today.

**Reply by JavierMNieto** [reply]
@ananthakumaran Wow, this is really nice, thanks. Really appreciate the work you're doing here. From the docs, this is exactly what I was imagining with the rollover and budget allocation calculations. I think putting the budget on its own page is a good call to not clutter up the expenses page.

I'll have to check it out more tomorrow to give some more feedback with my own budget.

**Reply by logeshpaul** [reply]
@ananthakumaran Great work!!

**Reply by JavierMNieto** [reply]
@ananthakumaran still fiddling around with it, but I am a little confused about the budget deficit calculation. I think it stems from this calculation [here](https://github.com/ananthakumaran/paisa/blob/5dbdf15ef527c1049d96ffc16e8f59268f87c76e/internal/server/budget.go#L106).

I noticed this includes forecasted transactions, so taking the last month (which could be 3 years in the future) does not seem right. Please correct me if I am wrong. I think maybe it is because forecasted income is not included? Along with budgeting, I also have periodic transactions for my income. Which brings me to another suggestion, which would be to have income presented on the budget screen somewhere to better visualize how budget compares to (expected) income.

**Reply by ananthakumaran** [reply]
> I think maybe it is because forecasted income is not included?

Yes, currently only expense postings are considered. Any future income postings etc. are not considered. I have been looking around to figure out how people do budgeting and one of the popular styles is [YNAB](https://www.ynab.com/the-four-rules/). Basically, you only budget the money you have in your checking account and if there is no deficit, then you can spend the money available under each category.

if we include the forecast income, then things become more complicated. Let's say you added an income transaction in the 3rd week of the month and the budget page may say all is okay, and when you try to spend some in the 2nd week, your checking account might not have enough balance.

Essentially, if we consider any future income, then we also need to convey the information that X amount will be available only after Y date, etc.

I am still thinking about it, feel free to point me to more literature about this. I will also do the implementation in stages. Might just go with the simple approach first and progressively add more capability to cover other use cases.

**Reply by dharapvj** [reply]
I am also getting foxed by the budget deficit amount. Since the budget page is shown for a given month, ideally, it should consider available income and budgeted items to show deficit for _that month_
